aten/src/ATen/native/README.md

```diff
@@ -384,6 +390,72 @@ will be automatically differentiated! This can be the case if the function imple
+to use in native_functions.yaml as it gives dispatcher information about backend and autograd support
+of the implementation.
+
+Here're steps to follow to decide the right dispatch keyword:
+
+1. Think about inference, does your kernel work for all backends?
+
+    - No: you're likely providing different kerenls for different backends, e.g. DispatchStub or
+      backend-depenent logics are used in the implementation.
+      Write a dispatch section and enumerate all supported backends and point them to the implementations.
+      ```
+      dispatch:
+        CPU: kernel_cpu
+        CUDA: kernel_cuda
+        QuantizedCPU: kernel_quantized_cpu
+      ```
+
+      [END] Now this op will be called in `CPU/CUDA/QuantizedCPU` backend inference!
+
+      Note if you want this op to support training as well, you're required to write a formula in
+      derivatives.yaml since your backend implementations don't support autograd.
+
+    - Yes: you're likely calling other `at::` ops in the implemetation. Go to step 2.
+2. Think about training, does your kernel support autograd? [check autograd support](#will-your-function-be-automatically-differentiable)
+    - Yes: in other words, you're providing a `Math` kernel which supports both inference and autograd.
+      To use autograd support for training, simply skip adding a dispatch section or write
+      ```
+      dispatch:
+        Math: kernel
+      ```
+
+      [END] This will allow this op to be correctly registered for both inference and training.
+
+    - Yes, but you still want to provide a numerically stable gradient formula instead of using autorgrad, write
+      ```
+      dispatch:
+        DefaultBackend: kernel
+      ```
+
+      [END] Now this op is only used in inference. Note that now you're required to add a autograd formula,
+      or it'll error out in forward pass when calling with a Tensor has requires_grad=True.
+
+    - No: ops in this category are mainly `_out` functions that don't have a derivative formula defined.
+      Write
+      ```
+      dispatch:
+        DefaultBackend: kernel
+      ```
+
+      [END] Now this op is only used in inference. Note that now you're required to add a autograd formula,
+      or it'll error out in forward pass when calling with a Tensor has requires_grad=True.
+
+      Note current plan on record for `_out` ops in this category is to replace `at::` with `at::native` in
+      the implementations and add dispatch section with device keywords instead.
+
+3. TODO: AutogradCPUOrCUDA
+
+Note that in native_functions.yaml you can mix using backend keywords and alias keywords above for one op:
+  - direct registration to backend always has higher precendence than alias
+  - DO NOT provide multiple alias keywords to the same op: alias keywords have precedence `DefaultBackend > Math`,
+    e.g. adding both `Math` and `DefaultBackend` kernels for one op will completely ignore `Math` kernel for
+    both inference and training.
```

Comment:
I thought we asserted on this case?

Reply:
Yup we have an assert in this case, adding it here just in case developers need to understand why a DefaultBackend is needed. 
Also just as a note here: currently we don't fail if users directly register to Math and DefaultBackend using a C++ API (this is unlikely but possible).